src/Morphir/Elm/Backend/Codec/DecoderGen.elm

```diff
@@ -0,0 +1,236 @@
+                    FunctionOrValue
+                        []
+                        ("decoder" ++ (name |> Name.toTitleCase))
+
+                _ ->
+                    Literal """Only string bool float
+                    and int primitives are supported"""
+
+        Record fields _ ->
+            let
+                mapFunc : Expression
+                mapFunc =
+                    fields
+                        |> List.length
+                        |> mapDecoderMapFunc
+
+                ctorFunc : Expression
+                ctorFunc =
+                    typeName
+                        |> Name.toTitleCase
+                        |> FunctionOrValue []
+
+                fieldDecoder : Name -> Field extra -> Expression
+                fieldDecoder _ (Field fieldName fieldType) =
+                    Application <|
+                        [ FunctionOrValue decoderModuleName "at" |> Utils.emptyRangeNode
+                        , ListExpr
+                            ((topLevelFieldNames
+                                |> List.map Literal
+                                |> List.map Utils.emptyRangeNode
+                             )
+                                ++ [ typeName |> Name.toCamelCase |> Literal |> Utils.emptyRangeNode
+                                   , fieldName |> Name.toCamelCase |> Literal |> Utils.emptyRangeNode
+                                   ]
+                            )
+                            |> Utils.emptyRangeNode
+                        , fieldType
+                            |> typeToDecoder (Name.fromString "") []
+                            |> Utils.emptyRangeNode
+                        ]
+
+                fieldDecoders : List Expression
+                fieldDecoders =
+                    fields
+                        |> List.map (fieldDecoder typeName)
+                        |> List.map Utils.emptyRangeNode
+                        |> List.map ParenthesizedExpression
+            in
+            Application <|
+                [ mapFunc |> Utils.emptyRangeNode
+                , ctorFunc |> Utils.emptyRangeNode
+                ]
+                    ++ (fieldDecoders |> List.map Utils.emptyRangeNode)
+
+        _ ->
+            Literal "Only reference and record types are supported"
+
+
+decoderModuleName : ModuleName
+decoderModuleName =
+    [ "D" ]
+
+
+mapDecoderMapFunc : Int -> Expression
+mapDecoderMapFunc n =
+    let
+        mapFuncName =
+            case n of
+                1 ->
+                    "map"
+
+                2 ->
+                    "map2"
+
+                3 ->
+                    "map3"
+
+                4 ->
+                    "map4"
+
+                5 ->
+                    "map5"
+
+                6 ->
+                    "map6"
+
+                7 ->
+                    "map7"
+
+                8 ->
+                    "map8"
+
+                _ ->
+                    "more than 8 fields cannot be mapped"
```

Comment:
This should be handled as an error. Currently it just generates invalid code. 

I also found a comment in elm/json that pointed me to https://package.elm-lang.org/packages/NoRedInk/elm-json-decode-pipeline/latest to support more than 8 fields.